lib/aws/ses.rb

```diff
@@ -21,11 +21,15 @@ def ses_client
       end
 
       def ses_client_options
-        region = Figaro.env.aws_ses_region
         opts = {}
-        opts[:region] = region if region.present?
+        opts[:region] = pick_region_from_pool if Figaro.env.aws_ses_region_pool.present?
         opts
       end
+
+      def pick_region_from_pool
+        pool = JSON.parse(Figaro.env.aws_ses_region_pool)
+        Pickup.new(pool).pick
```

Comment:
Was there a particular reason for choosing this gem as opposed to rolling our own 1-method solution? Maybe something like this, which I found doing a quick search:
```ruby
def choose_weighted(weighted)
  sum = weighted.inject(0) do |sum, item_and_weight|
    sum += item_and_weight[1]
  end
  target = rand(sum)
  weighted.each do |item, weight|
    return item if target <= weight
    target -= weight
  end
end
```

Reply:
Or even simpler:
```ruby
pool = JSON.parse(Figaro.env.aws_ses_region_pool)
regions = []
 
pool.map do |region, percentage|
  percentage.times { regions << region }
end

regions.sample
```
Reference: http://www.rubyguides.com/2016/05/weighted-random-numbers/

Reply:
The latter is the fastest option. Given our email performance issues, I think we'd want to choose the fastest solution. To make it even faster, we can make the regions array a constant in an initializer, and then all we have to do is randomly pick from it.

Reply:
As always, I run benchmarks when comparing solutions:
```
sample:  7121661.2 i/s
random_weighted:   725578.4 i/s - 9.82x  slower
pickup:    80840.0 i/s - 88.10x  slower
```
As you can see, using the Pickup gem is a whopping 88 times slower than randomly sampling an array.

Here is the script I used inside the Rails console:
```ruby
require 'benchmark/ips'

POOL = JSON.parse('{ "us-west-2": 5, "us-east-1": 95 }')
REGIONS = POOL.each_with_object([]) do |(region, percentage), regions|
  percentage.times { regions << region }
  region
end

def random_weighted(weighted)
  max    = sum_of_weights(weighted)
  target = rand(1..max)

  weighted.each do |item, weight|
    return item if target <= weight
    target -= weight
  end
end

def sum_of_weights(weighted)
  weighted.inject(0) { |sum, (item, weight)| sum + weight }
end

Benchmark.ips do |x|
  x.report("sample") { REGIONS.sample }

  x.report("random_weighted") { random_weighted(POOL) }

  x.report('pickup') { Pickup.new(POOL).pick }

  x.compare!
end
```

Reply:
I used the gem to keep the code in the SES client and its tests simple, but I can pull in a function to select a region on our own.

I can also move the region array into a initializer, but I think it makes the code a bit trickier to follow for a pretty small gain.